io-sim/src/Control/Monad/IOSim/Internal.hs

```diff
@@ -634,7 +647,7 @@ schedule thread@Thread{
         -- the main thread is done, so we're done
         -- even if other threads are still running
         return $ Trace time tid tlbl EventThreadFinished
-               $ TraceMainReturn time x (Map.keys threads)
+               $ TraceMainReturn time x (labeledThreads threads)
```

Comment:
Previously returning the keys _I think_ would not retain the whole map, but this will. But if we make `labeledThreads` a bit stricter then it will not retain the threads mapping.